### ISSUE TITLE
How to update data of a node?

### ISSUE THREAD

**Issue by xuanngo2001** [bumbeishvili/org-chart#175]
I implemented a popup to allow users to change data in the node, e.g. firstname.
When users clicked on a node, a popup appear and they can change the firstname.
On save, I use javascript to change the firstname of the node but the problem is org-chart still hold the old firstname. If the users clicked again on the node, the old value will be shown.

How can I update the data of a node?

**Comment by bumbeishvili** [comment]
On change you can pass updated data and rerender the graph

`chart.date(updatedData).render()`

**Comment by xuanngo2001** [comment]
Does that imply that I have to pass the **whole** new data back to chart and it will render() the **whole** graph? Wouldn't I lost the state of the graph, e.g. selected node, nodes that are expanded/collapsed, etc?

**Comment by bumbeishvili** [comment]
Yes , and no, you won't lose state . You can check the angular example and you'll see how the updated data is passed and how the graph reacts

**Comment by Shadowfita** [comment]
Hi @bumbeishvili ,

Thank you very much for your work on this chart. It's a very useful piece of software.

I've just started working with re-rendering the chart on data update, however, I am losing state upon re-render, and all nodes collapse. Is there a way to avoid this?

Thanks heaps.

**Comment by bumbeishvili** [comment]
@Shadowfita  not sure why you are losing state. Data is being mutated in the OrgChart itself (it attaches its state), can it be because you are passing a new data?

**Comment by Shadowfita** [comment]
> @Shadowfita not sure why you are losing state. Data is being mutated in the OrgChart itself (it attaches its state), can it be because you are passing a new data?

Oh that is exactly what it is. Complete oversight, haha. Is it possible to pass updated data (added single new node to dataset) without losing the rest of the state?

**Comment by bumbeishvili** [comment]
It's possible to add nodes and pass the new data, what I mean is that the data attached to old nodes should not be cleared. A.K.A. old nodes should not be constructed again from the scratch

You can check React example how it handles similar scenario

https://stackblitz.com/edit/d3-org-chart-react-integration-hooks?file=index.js

**Comment by Shadowfita** [comment]
Thanks @bumbeishvili , I've gone ahead and made use of the chart.add/removeNode function instead of updating dataset directly and removed re-render on data update.
